File: skellig-feature/src/main/java/org/skellig/feature/TestStep.java
package org.skellig.feature;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.stream.Collectors;

public class TestStep {

    private String name;
    private Map<String, String> parameters;

    protected TestStep(String name, Map<String, String> parameters) {
        this.name = name;
        this.parameters = parameters;
    }

    public String getName() {
        return name;
    }

    public Optional<Map<String, String>> getParameters() {
        return Optional.ofNullable(parameters == null ? null : Collections.unmodifiableMap(parameters));
    }

    public static class Builder {
        private String name;
        private Map<String, String> parameters;

        public Builder withName(String name) {
            this.name = name.trim();
            return this;
        }

        public Builder withParameters(Map<String, String> parameters) {
            this.parameters = parameters;
            return this;
        }

        public Builder withParameter(String name, String value) {
            if (parameters == null) {
                parameters = new HashMap<>();
            }
            this.parameters.put(name.trim(), value.trim());
            return this;
        }

        public TestStep build() {
            return new TestStep(name, parameters);
        }

        public TestStep buildAndApplyTestData(Map<String, String> testData) {
            Map<String, String> newParameters = getParametersWithAppliedTestData(testData);
            return new TestStep(ParametersUtils.replaceParametersIfFound(name, testData), newParameters);
        }

        private Map<String, String> getParametersWithAppliedTestData(Map<String, String> dataRow) {
            return parameters == null ? null :
                    parameters.entrySet().stream()
                            .collect(Collectors.toMap(Map.Entry::getKey,
                                    entry -> ParametersUtils.replaceParametersIfFound(entry.getValue(), dataRow)));
        }
    }
}

File: skellig-feature/src/main/kotlin/org/skellig/feature/TestStep.kt
package org.skellig.feature

class TestStep private constructor(
    val path: String,
    val name: String,
    val parameters: Map<String, Any?>?
) : SkelligTestEntity {

    override fun getEntityName(): String = name

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TestStep

        return path == other.path
    }

    override fun hashCode(): Int = path.hashCode()

    class Builder {
        private var position = 0
        private var parent: String? = null
        private var name: String? = null
        private var parameters: MutableMap<String, String?>? = null

        fun withName(name: String) = apply { this.name = name.trim { it <= ' ' } }

        fun withParameters(parameters: MutableMap<String, String?>?) = apply { this.parameters = parameters }

        fun withParent(parent: String) = apply { this.parent = parent }

        fun withPosition(position: Int) = apply { this.position = position }

        fun build(): TestStep {
            return TestStep(createPath(), name!!, parameters)
        }

        fun buildAndApplyTestData(testData: Map<String, String>): TestStep {
            val newParameters = getParametersWithAppliedTestData(testData)
            return TestStep(createPath(), ParametersUtils.replaceParametersIfFound(name!!, testData), newParameters)
        }

        private fun createPath() = "$parent:$name:$position"

        private fun getParametersWithAppliedTestData(dataRow: Map<String, String>): Map<String, Any?>? {
            return parameters
                ?.map { it.key to it.value?.let { v -> ParametersUtils.replaceParametersIfFound(v, dataRow) } }
                ?.toMap()
        }
    }
}